gobblin-core/src/main/java/gobblin/source/extractor/extract/google/webmaster/GoogleWebmasterDataFetcherImpl.java

```diff
@@ -0,0 +1,291 @@
+            pagesNoRoots.size(), pagesRoots.size()));
+      }
+    }
+
+    return new ImmutableTriple<>(pagesNoRoots, pagesRoots, entireSize);
+  }
+
+  /**
+   * This method gives you all pages given the page prefixes/keywords/root paths.
+   *
+   * The detailed steps are as follows:
+   *
+   * 1. For each page prefix, construct a filter group with two filters. One for country; the other for containing that page prefix.
+   * 2. Request for all pages with the filter group. If the count is 5000, go to step 3; otherwise, end of the process, we've already got all pages.
+   * 3. Expand current prefix by appending a char from the list [a-z,0-9,%,_,-,/].
+   * 4. Perform another request for each new prefix at last step. If the number of pages is 5000, repeat steps 3~4; otherwise, end of the process, we've already got all pages.
+   * 5. Union all pages as the return value
+   *
+   * @param date
+   * @param dimensions requested dimensions
+   * @param pagePrefixes a pagePrefix must end with "/"
+   * @return all pages given the page prefixes/keywords, or page root paths.
+   * @throws IOException
+   */
+  private Set<String> pagesWithRootPaths(String date, List<Dimension> dimensions, ApiDimensionFilter countryFilter,
+      Set<String> pagePrefixes) throws IOException {
+
+    for (String prefix : pagePrefixes) {
+      Preconditions.checkArgument(prefix.endsWith("/"), "Starting prefix must end with '/'");
+    }
+    Country countryString = countryFilterToEnum(countryFilter);
+
+    List<ApiDimensionFilter> filters = new LinkedList<>();
+    filters.add(countryFilter);
+    filters.add(null); //placeholder for page filter
+
+    Set<String> uniquePages = new HashSet<>();
+    Deque<String> toProcess = new LinkedList<>(pagePrefixes);
+    while (!toProcess.isEmpty()) {
+      String prefix = toProcess.remove();
+      LOG.info("Current page prefix is " + prefix);
+      filters.set(1, GoogleWebmasterFilter.pageFilter(GoogleWebmasterFilter.FilterOperator.CONTAINS, prefix));
+      List<String> pages =
+          _client.getPages(_siteProperty, date, countryString, GoogleWebmasterClient.API_ROW_LIMIT, dimensions, filters,
+              0);
+
+      if (pages.size() == GoogleWebmasterClient.API_ROW_LIMIT) {
+        //If the number of pages is at the LIMIT, we need to create sub-tasks and redo query
+        LOG.info(String.format(
+            "Number of pages fetched for prefix %s reaches the MAX request limit %d. Expanding the prefix...", prefix,
+            GoogleWebmasterClient.API_ROW_LIMIT));
+        //The page prefix is case insensitive, A-Z is not necessary.
+        for (char c = 'a'; c <= 'z'; ++c) {
+          toProcess.add(prefix + c);
+        }
+        for (int num = 0; num <= 9; ++num) {
+          toProcess.add(prefix + num);
+        }
+        toProcess.add(prefix + "%");
+        toProcess.add(prefix + "_");
+        toProcess.add(prefix + "-");
+        toProcess.add(prefix + "/");
```

Comment:
This is not an exhaustive list. In reality, only prefix + "%" is enough. Checking for the exhaustive list is too expensive, so I only checked four most likely characters. I may need to think about a different way and rewrite the whole piece here.